pdr_backend/lake/table_bronze_pdr_slots.py

```diff
@@ -0,0 +1,203 @@
+def _process_truevals(
+    dfs: Dict[str, pl.DataFrame], ppss: PPSS
+) -> Dict[str, pl.DataFrame]:
+    """
+    Perform post-fetch processing on the data
+    """
+    # get truevals within the update
+    truevals_df, _ = pick_df_and_ids_on_period(
+        target=dfs["pdr_truevals"],
+        start_timestamp=ppss.lake_ss.st_timestamp,
+        finish_timestamp=ppss.lake_ss.fin_timestamp,
+    )
+
+    # get ref to bronze_slots
+    slots_df = dfs[bronze_pdr_slots_table_name]
+
+    # add columns that are going to be populated
+    slots_df = slots_df.with_columns(last_event_timestamp=pl.lit(None))
+    slots_df = slots_df.with_columns(trueval=pl.lit(None))
+
+    initial_schema_keys = slots_df.schema.keys()
+
+    # update only the ones within this time range
+    slots_df = (
+        slots_df.join(truevals_df, left_on="ID", right_on="ID", how="left")
+        .with_columns(
+            [
+                pl.col("trueval").fill_null(pl.col("trueval")),
+                pl.col("timestamp_right").fill_null(pl.col("last_event_timestamp")),
+            ]
+        )
+        .drop(["trueval", "last_event_timestamp"])
+        .rename(
+            {
+                "trueval_right": "trueval",
+                "timestamp_right": "last_event_timestamp",
+            }
+        )
+        .select(initial_schema_keys)
+    )
+
+    # update dfs
+    dfs[bronze_pdr_slots_table_name] = slots_df
+    return dfs
+
+
+def _process_predictions(
+    dfs: Dict[str, pl.DataFrame], ppss: PPSS
+) -> Dict[str, pl.DataFrame]:
+    """
+    @description
+        Perform post-fetch processing on the data
+        1. Find predictions( within the update
+
+    """
+    # get predictions within the update
+    predictions_df, _ = pick_df_and_ids_on_period(
+        target=dfs["pdr_predictions"],
+        start_timestamp=ppss.lake_ss.st_timestamp,
+        finish_timestamp=ppss.lake_ss.fin_timestamp,
+    )
+
+    # get existing bronze_predictions we'll be updating
+    slots_df = dfs[bronze_pdr_slots_table_name]
+
+    # add columns that are going to be populated
+    slots_df = slots_df.with_columns(pair=pl.lit(None))
+    slots_df = slots_df.with_columns(source=pl.lit(None))
+    slots_df = slots_df.with_columns(timeframe=pl.lit(None))
+
+    initial_schema_keys = slots_df.schema.keys()
+
+    # do work to join from pdr_payout onto bronze_pdr_predictions
+    slots_df = (
+        slots_df.join(predictions_df, on=["contract"], how="left")
+        .with_columns(
+            [
+                pl.col("pair").fill_null(pl.col("pair")),
+                pl.col("source").fill_null(pl.col("source")),
+                pl.col("timeframe").fill_null(pl.col("timeframe")),
+            ]
+        )
+        .drop(["pair", "source", "timeframe"])
+        .rename(
+            {
+                "pair_right": "pair",
+                "source_right": "source",
+                "timeframe_right": "timeframe",
+            }
+        )
+        .select(initial_schema_keys)
```

Comment:
**[Columns Missing]**
We want to track all predictions that happened to the slot.

Please create a column named `predictions` with a blob-type parameter:
```
pseudo schema: {
...,
predictions: list[str] #or a comma delineated str
}
```

The key thing we want to append to this, are all the prediction_ids from all the users that have joined this epoch. If the user is already in here, it should not be added again.
```
- prediction_id
```

Now, when we got to enrich/explode this data, we can join w/ `bronze_prediction` tables.

Reply:
I don't think this is necessary because we already have the contract and slot fields inside the bronze_slots and we can get the predictions based on that. Instead of prediction id just get by same contract and slot, then no need fot the extra logic to keep the list of prediction ids

Reply:
Ok, we're not going to fix this now and will fix this in the future when implementing the `pdr_contracts` table